nedap/speced.def
# ClojureScript compatibility + Primitives support + Type inference

# Brief

utils.spec is approaching drinking age!

### ClojureScript compatibility

For all practical matters, it:

* fully works,
* leverages the same syntax(es) as JVM utils.spec,
* enjoys full test coverage,
* and is CI'd right now.

### Primitives support

You now can add `^long` / `^double` ([etc](https://clojure.org/reference/java_interop#TypeAliases)) annotations (in JVM) and `^string` / `^number` / `^boolean` annotations (JS; only those three). Both will result in more performant code than `^string?` / `^String` / `^js/String`, while enjoying the usual spec checking.

* It's only for defn: it doesn't work for protocols in JVM clojure.core, so it doesn't here either.
  * It _could_ work in cljs.core, but I don't want to invite people to write non-portable code that doesn't even bring a perf benefit
* I'm tempted to start recommending prefering the `^long` annotation whenever possible
  * Yields faster code
  * the compiler takes care of coercion (when sensible), so it's quite unrisky to use

### Type inference

Now you can write a  `^string? x` signature, and the right type hint (`String` / `string` for JVM/JS) will be emitted

* Now your code will be effortlessly faster and warn-free (as per Eastwood/formatting-stack)
* Every single predicate from clojure.core has a utils.spec pred->class mapping
  * The mechanism isn't particularly extensible right now, but you can alter-var-root it if you wish.
* Note: there's only type inferences for functions (`^string?`) and primitives (`^long`) but not for keywords (`::my/spec`)
  * Certainly can/should be implemented at some point

# Review plan

The three changes were quite interrelated so I crafted a large PR, sorry. The history is clean though and most commits can be reviewed in isolation.

You also can just browse / play with the test suite to be convinced of the quality.

The single most important test is [defn](https://github.com/nedap/utils.spec/blob/9c57894f07d6d946e9b076e3861acf9af6fd5a79/test/unit/nedap/utils/speced/defn.cljc)'s.

# QA plan

* For existing JVM users: you can install 0.7.1-alpha2 and check if your test suite still passes. It [does](https://circleci.com/gh/nedap/pep-key/3705) in Key.
  * (that build fails unrelatedly; the suite passes)

* For new CLJS users: you could add 0.7.1-alpha2, and use `speced/defn` for at least one existing defn, adding some specs.
  * Verify that the spec checking works
  * Verify that no compilation warnings are emitted
  * Verify that the production build size doesn't skyrocket
    * It should have essentially no cost if correctly setting up [:elide-asserts](https://clojurescript.org/reference/compiler-options#elide-asserts), plus minification.

## Minor TODOs

Reflected in https://github.com/nedap/utils.spec/issues/57, to be fixed asap
